### DISCUSSION TITLE
Modular code

### DISCUSSION THREAD

**Discussion by owennewo** [zalo/CascadeStudio#78] [Ideas]
Are there any ways to modulize your code.  e.g. to load functions, classes or code from from git repos?

I had a go at trying to load remote code (from a github gist) - I could load it but couldn't work out how to render it

```
function loadFileAsync(url) {
    return new Promise(function (resolve, reject) {
        let xhr = new XMLHttpRequest();
        xhr.open("GET", url);
        xhr.onload = function () {
            if (this.status >= 200 && this.status < 300) {
                console.log(xhr.response);
                resolve(xhr.response);
            } else {
                reject({
                    status: this.status,
                    statusText: xhr.statusText
                });
            }
        };
        xhr.onerror = function () {
            console.log("load error");
            reject({
                status: this.status,
                statusText: xhr.statusText
                
            });
        };
        xhr.send();
    });
}

async function loadFileSync(url) {
    let result = await loadFileAsync(url);
    console.log(result);
    return result;
}

console.log(loadFileSync("https://gist.githubusercontent.com/owennewo/28d927d43f7febfd7bcd6d0939ed61c6/raw/dfd0a717774bfc0b293800011f38368e81788eac/gistfile1.txt"));

```

The above prints the remote text but also get an
```
Line 11: Uncaught DataCloneError: Failed to execute 'postMessage' on 'DedicatedWorkerGlobalScope': # could not be cloned.
```

Oh and another thing, is it possible to switch on ES2015 language support. Things like Promise and foreach work but show up as soft errors that can be ignored.

This project is awesome BTW!!

**Comment by Irev-Dev** [comment]
I don't have an answer for you @owennewo, but I'm curious what the use case is?
Are you trying to reuse code? or is it that you want your project to have a bit more persistence?

I've been thinking about a github integration that would render a model based on a repo, but I haven't got any further than thinking about it. 🤷‍♂️.

**Comment by owennewo** [comment]
For me it's about reuse.   But it could also be about sharing a function/complex component.

I make a lot of shims between different sized motors and different sized angle sensors. They are similar to each other but not the same.  The base part e.g support is often identical, but then i do a difference to carve out holes to match the dimensions of a specific motor.

At the moment there is a fair bit of copy/paste involved.
